### PR TITLE
[Core] Fix opencensus protobuf bug

### LINKED ISSUE #31358: [Jobs]  Couldn't build proto file into descriptor pool! exception after installing ray in project
### Couldn't build proto file into descriptor pool! exception after installing ray in project

Hi, I have an existing working project in which I added the ray[default] dependency. Now the following has started coming.

```
from opentelemetry.exporter.opencensus.trace_exporter import OpenCensusSpanExporter
  File "/tmp/ray/session_2022-12-23_07-50-00_227742_8/runtime_resources/pip/513edf5d83ec05c69a73be1be828c66f252e0b48/virtualenv/lib/python3.7/site-packages/opentelemetry/exporter/opencensus/trace_exporter/__init__.py", line 21, in <module>
    from opencensus.proto.agent.trace.v1 import (
  File "/tmp/ray/session_2022-12-23_07-50-00_227742_8/runtime_resources/pip/513edf5d83ec05c69a73be1be828c66f252e0b48/virtualenv/lib/python3.7/site-packages/opencensus/proto/agent/trace/v1/trace_service_pb2.py", line 16, in <module>
    from opencensus.proto.resource.v1 import resource_pb2 as opencensus_dot_proto_dot_resource_dot_v1_dot_resource__pb2
  File "/tmp/ray/session_2022-12-23_07-50-00_227742_8/runtime_resources/pip/513edf5d83ec05c69a73be1be828c66f252e0b48/virtualenv/lib/python3.7/site-packages/opencensus/proto/resource/v1/resource_pb2.py", line 22, in <module>
    serialized_pb=_b('\n+opencensus/proto/resource/v1/resource.proto\x12\x1copencensus.proto.resource.v1\"\x8b\x01\n\x08Resource\x12\x0c\n\x04type\x18\x01 \x01(\t\x12\x42\n\x06labels\x18\x02 \x03(\x0b\x32\x32.opencensus.proto.resource.v1.Resource.LabelsEntry\x1a-\n\x0bLabelsEntry\x12\x0b\n\x03key\x18\x01 \x01(\t\x12\r\n\x05value\x18\x02 \x01(\t:\x02\x38\x01\x42\x98\x01\n\x1fio.opencensus.proto.resource.v1B\rResourceProtoP\x01ZEgithub.com/census-instrumentation/opencensus-proto/gen-go/resource/v1\xea\x02\x1cOpenCensus.Proto.Resource.V1b\x06proto3')
  File "/home/ray/anaconda3/lib/python3.7/site-packages/google/protobuf/descriptor.py", line 982, in __new__
    return _message.default_pool.AddSerializedFile(serialized_pb)
TypeError: Couldn't build proto file into descriptor pool!
Invalid proto descriptor for file "opencensus/proto/resource/v1/resource.proto":
  opencensus/proto/resource/v1/resource.proto: A file with this name is already in the pool.
```

### Versions / Dependencies

opentelemetry-exporter-opencensus = 0.17b0
ray = 2.2.0

### Reproduction script

test.py

```
from opentelemetry.exporter.opencensus.trace_exporter import OpenCensusSpanExporter
from ray.job_submission import JobSubmissionClient, JobStatus
```

pyproject.toml

```
[tool.poetry]
name = "testing"
version = "0.1.0"
description = "Testing Library"
authors = ["Engineering <navinder_brar@yahoo.com>"]

[[tool.poetry.source]]
name = 'pypi-public'
url = "https://pypi.org/simple/"

[tool.poetry.dependencies]
python = "^3.7"
ray = {extras = ["default"], version = "^2.2.0"}
opentelemetry-api = "^0.17b0"
opentelemetry-sdk = "^0.17b0"
opentelemetry-instrumentation-grpc = "^0.17b0"
opentelemetry-instrumentation = "^0.17b0"
opentelemetry-exporter-opencensus = "^0.17b0"
opentelemetry-exporter-prometheus = "^0.17b0"
protobuf = "3.20.3"

[build-system]
requires = ["poetry>=0.12", "pip>=20.2"]
build-backend = "poetry.masonry.api"
```

After installing  dependencies via poetry install -> run test.py

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Here's how the conflict happens.

1. We pull the opencensus-proto 0.3.0 for our bazel build.
2. opencensus-proto 0.3.0 generated python code (metric_pb2 and resource_pb2), which is used by `reporter.proto`. 
3. We copy all python protobuf to python/ray/core/generated.
4. Now reporter_pb2 is available from python/ray/core/generated.

reporter_pb2 is not supposed to be imported from core workers (it is only for dashboard agent). However, this was imported (accidentally) from gcs_pubsub.py because pubsub.proto requires this module (which was unnecessary). 

As a result, reporter_pb2 was imported to the core worker, which subsequently imports metric_pb2 and resource_pb2. When we import the opencensus proto using open telemetry, we import resource_pb2 again, which causes the conflict.

Theoretically, when the same package is imported (which is our case), protobuf is supposed to handle it. (I guess) This wasn't the case because our cpp version uses opencensus-proto 0.3.0, and the python only supports opencensus-proto 0.1.0 (omg https://pypi.org/project/opencensus-proto/#history).

There are 4 potential solutions. 

1. [**current PR**] Make sure to not import reporter_pb2 from the core worker. It is the easiest fix and have the best compatibility story. Since we don't import reporter_pb2 (which is only for dashboard agent), there's no concern of conflict. Since we have unit tests now, we can avoid the future regressions.
2. Use downloaded opencensus-proto for all python generated code (instead of cpp one we built). This requires us to introduce opencensus-proto 0.1.0 as a dependency, so I avoided this approach. 
3. Downgrade cpp opencensus-proto version to 0.1.0. This is not a very clean solution. What if opencensus releases 0.2.0?
4. Copy & paste opencensus-proto and build our own proto with different names. This can potentially have compatibility issue with opencensus library, but it is unlikely (cuz the project doesn't seem to be maintained). This is the fallback option if the current approach doesn't work. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/31358

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/test_utils.py [hunk 1]
```diff
@@ -50,7 +50,6 @@
 )
 from ray.scripts.scripts import main as ray_main
 from ray.util.queue import Empty, Queue, _QueueActor
-from ray.experimental.state.state_manager import StateDataSourceClient
```
Comment (rkooo567) [inline python/ray/_private/test_utils.py:53]: state manager imports reporter_agent which breaks the test.

Comment (rkooo567) [reply python/ray/_private/test_utils.py:53]: so we should move it away from the test util (which is how it was supposed to be ...)

File: src/ray/protobuf/pubsub.proto [hunk 2]
```diff
@@ -21,7 +21,6 @@ import "src/ray/protobuf/common.proto";
 import "src/ray/protobuf/dependency.proto";
 import "src/ray/protobuf/gcs.proto";
 import "src/ray/protobuf/logging.proto";
-import "src/ray/protobuf/reporter.proto";
```
Comment (rkooo567) [inline src/ray/protobuf/pubsub.proto:24]: this is the critical fix

**Comment by rkooo567** [comment]
@scv119 I am 80% sure this will pass all builds, but let's see...

**Comment by rickyyx** [comment]
Looks like some tests are still failing?

> Theoretically, when the same package is imported (which is our case), protobuf is supposed to handle it. (I guess) This wasn't the case because our cpp version uses opencensus-proto 0.3.0, and the python only supports opencensus-proto 0.1.0 (omg https://pypi.org/project/opencensus-proto/#history).

This is really really old....Looks like it's not maintained at all. Do we have any plans to migrate away from it if so?

**Review by fishbone — APPROVED** [review]

**Comment by scv119** [comment]
@rkooo567 is the long term solution to put related protobuf into a different namespace so it will never conflict?

**Comment by rkooo567** [comment]
@scv119 That's solution 4. I think the current solution is good enough in the long term (since we have a regression test now). If we ever have to import reporter.proto to core worker, solution 2 or 4 both will work.

File: python/ray/_private/test_utils.py (hunk 1 above)
Comment (rickyyx) [reply python/ray/_private/test_utils.py:53]: hmm, why this was a necessary change? Does this mean we couldn't import state_manager from anywhere?

Comment (rkooo567) [reply python/ray/_private/test_utils.py:53]: test_metrics -> test_utils -> state_manager -> reporter_pb2

And since it is a test, the driver is affected.

It is okay to import this. We should just make sure this is not imported to the core worker code (which makes sense). For most of tests it is fine because we don't import opencensus proto.

**Comment by rkooo567** [comment]
If anyone reorg proto or do something that breaks it, the test will fail, so we can detect them. If this ever should happen, we can choose option 4 (I will add TODO). 

The only edge case possible is an end user imports reporter_pb2 (private) to the driver and uses opencensus proto at the same time. Tbh, I don’t think there will be anyone who will do this... Note that this issue has existed for 3 years, and this was the first issue ever raised.

**Review by rickyyx — APPROVED** [review]
<div><br/>via <a href="https://giphy.com/martinonassis/">Martin Onassis</a> on <a href="https://giphy.com/gifs/brain-laser-15-folds-l0HlEEwgZfgqfH70c">GIPHY</a></div>
